coalib/bearlib/abstractions/Linter.py

```diff
@@ -0,0 +1,680 @@
+from contextlib import contextmanager
+import inspect
+from itertools import compress
+import re
+import shutil
+from subprocess import check_call, CalledProcessError, DEVNULL
+from types import MappingProxyType
+
+from coalib.bears.LocalBear import LocalBear
+from coalib.misc.ContextManagers import make_temp
+from coalib.misc.Decorators import assert_right_type, enforce_signature
+from coalib.misc.Future import partialmethod
+from coalib.misc.Shell import run_shell_command
+from coalib.results.Diff import Diff
+from coalib.results.Result import Result
+from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
+from coalib.settings.FunctionMetadata import FunctionMetadata
+
+
+def _prepare_options(options):
+    """
+    Prepares options for ``Linter`` for a given options dict in-place.
+
+    :param options:
+        The options dict that contains user/developer inputs.
+    """
+    allowed_options = {"executable",
+                       "output_format",
+                       "use_stdin",
+                       "use_stdout",
+                       "use_stderr",
+                       "config_suffix",
+                       "prerequisite_check_command"}
+
+    if not options["use_stdout"] and not options["use_stderr"]:
+        raise ValueError("No output streams provided at all.")
+
+    if options["output_format"] == "corrected":
+        if "diff_severity" in options:
+            if options["diff_severity"] not in RESULT_SEVERITY.reverse:
+                raise TypeError("Invalid value for `diff_severity`: " +
+                                repr(options["diff_severity"]))
+
+        if "diff_message" in options:
+            assert_right_type(options["diff_message"], str, "diff_message")
+
+        allowed_options |= {"diff_severity", "diff_message"}
+    elif options["output_format"] == "regex":
+        if "output_regex" not in options:
+            raise ValueError("No `output_regex` specified.")
+
+        options["output_regex"] = re.compile(options["output_regex"])
+
+        # Don't setup severity_map if one is provided by user or if it's not
+        # used inside the output_regex. If one is manually provided but not
+        # used in the output_regex, throw an exception.
+        if "severity_map" in options:
+            if "severity" not in options["output_regex"].groupindex:
+                raise ValueError("Provided `severity_map` but named group "
+                                 "`severity` is not used in `output_regex`.")
+            assert_right_type(options["severity_map"], dict, "severity_map")
+
+            for key, value in options["severity_map"].items():
+                try:
+                    assert_right_type(key, str, "<severity_map dict-key>")
+                except TypeError:
+                    raise TypeError("The key " + repr(key) + " inside given "
+                                    "severity-map is no string.")
```

Comment:
Because the error message from `assert_right_type` would be misleading then, as a key inside the severity_map is no parameter^^